### PR TITLE
feat: added CBR support

### PR DESCRIPTION
**Pull request by Khuzaima05**
### Description

This PR adds cbr support in [terraform-ibm-base-ocp-vpc](https://github.com/terraform-ibm-modules/terraform-ibm-base-ocp-vpc) module.

issue : https://github.ibm.com/orgs/GoldenEye/projects/3/views/8?filterQuery=assignee%3A%40me&pane=issue&itemId=430995

### Release required?

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [x] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

##### Release notes content

added support to create CBR rules using new input cbr_rules

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -646,3 +646,34 @@ module "attach_sg_to_registry_vpe" {
...
+  resources = [{
+    attributes = [
+      {
+        name     = "accountId"
+        value    = var.cbr_rules[count.index].account_id
+        operator = "stringEquals"
+      },
+      {
+        name     = "serviceInstance"
+        value    = ibm_container_vpc_cluster.cluster[0].crn
```
Comment (ocofaigh) [inline main.tf:669]: need to be careful here.. if the user provisioned with `var.ignore_worker_pool_size_changes` set to true, the resource will be `ibm_container_vpc_cluster.autoscaling_cluster[0].crn` instead.

File: main.tf [hunk 2]
```diff
@@ -646,3 +646,34 @@ module "attach_sg_to_registry_vpe" {
...
+        operator = "stringEquals"
+      },
+      {
+        name     = "serviceName"
+        value    = "Red Hat OpenShift on IBM Cloud"
```
Comment (ocofaigh) [inline main.tf:674]: Is that actually a valid service name @Ak-sky ?

**Review by ocofaigh — CHANGES_REQUESTED** [review]
I left some comments - @Ak-sky can you help guide here?
I think maybe we could add cbr into the fscloud example too, and update that so we allow traffic from the VPC and from the schematics IPs?

File: main.tf (hunk 2 above)
Comment (Ak-sky) [reply main.tf:674]: Correct service name is `containers-kubernetes`

File: main.tf (hunk 1 above)
Comment (Ak-sky) [inline main.tf:669]: Also the value should be cluster id not the crn

**Comment by Ak-sky** [comment]
> I left some comments - @Ak-sky can you help guide here? I think maybe we could add cbr into the fscloud example too, and update that so we allow traffic from the VPC and from the schematics IPs?

@Khuzaima05, as @ocofaigh mentioned, CBR has be added to fscloud module and in fscloud example also.

**Comment by Ak-sky** [comment]
Test is failing beacuse `The IKS service does not support service references in rules on "Cluster control plane" API`, [slack discussion](https://ibm-cloudplatform.slack.com/archives/C02EDELK7GB/p1709655055003959?thread_ts=1709654031.163439&cid=C02EDELK7GB).

File: examples/advanced/main.tf [hunk 3]
```diff
@@ -175,6 +212,34 @@ module "ocp_base" {
...
+        }, {
+        attributes = [
+          {
+            "name" : "endpointType",
+            "value" : "public"
+          },
+          {
+            name  = "networkZoneId"
+            value = module.cbr_zone_schematics.zone_id
+        }]
```
Comment (Ak-sky) [inline examples/advanced/main.tf:239]: Same code needs to be added for the fscloud example (schematics zone).

File: examples/advanced/main.tf [hunk 4]
```diff
@@ -175,6 +212,34 @@ module "ocp_base" {
     instance_id = module.kp_all_inclusive.kms_guid
     crk_id      = module.kp_all_inclusive.keys["${local.key_ring}.${local.cluster_key}"].key_id
   }
+  cbr_rules = [
+    {
+      description      = "${var.prefix}-event notification access only from vpc"
```
Comment (Ak-sky) [inline examples/advanced/main.tf:217]: Description needs correction.

File: examples/advanced/main.tf (hunk 3 above)
Comment (Ak-sky) [reply examples/advanced/main.tf:239]: Also to apply correct CBR rule here, you ned to pass `operations` attribute value in CBR rule.

```
operations = [{
    api_types = [
      {
        "api_type_id": "crn:v1:bluemix:public:containers-kubernetes::::api-type:management"
      }
    ]
  }]
```

**Comment by Ak-sky** [comment]
Schematics Apply job was successful but got timed out
image

```
=== NAME  TestAdvancedExampleInSchematics
    tests.go:140: 
        	Error Trace:	/go/pkg/mod/github.com/terraform-ibm-modules/ibmcloud-terratest-wrapper@v1.41.1/testschematic/tests.go:140
        	            				/__w/terraform-ibm-base-ocp-vpc/terraform-ibm-base-ocp-vpc/tests/pr_test.go:101
        	Error:      	Received unexpected error:
        	            	time exceeded waiting for schematic job to finish
        	Test:       	TestAdvancedExampleInSchematics
        	Messages:   	error waiting for APPLY to finish - 
```

**Comment by ocofaigh** [comment]
@Ak-sky you will need to make this longer: `WaitJobCompleteMinutes: 60,` <- it can take more than 60mins to deploy the advanced example

**Comment by Ak-sky** [comment]
Destroy falied due to `reasons='[PROTECTED_RESOURCE_ERR: Key is protecting one or more cloud resources`

```
 2024/10/27 06:58:20 Terraform destroy | ibm_is_vpc.vpc: Still destroying... [id=r050-41199242-3281-4d4d-b146-301fb020fbe3, 20s elapsed]
 2024/10/27 06:58:23 Terraform destroy | ibm_is_vpc.vpc: Destruction complete after 22s
 2024/10/27 06:58:23 Terraform destroy | 
 2024/10/27 06:58:23 Terraform destroy | Error: [ERROR] Error while deleting: kp.Error: correlation_id='5edcc499-2465-4a70-afbe-8e4ebc46484c', msg='Conflict: Key could not be deleted: Please see `reasons` for more details (PROTECTED_RESOURCE_ERR)', reasons='[PROTECTED_RESOURCE_ERR: Key is protecting one or more cloud resources - FOR_MORE_INFO_REFER: https://cloud.ibm.com/docs/key-protect?topic=key-protect-troubleshooting#unable-to-delete-keys]'. The key has the following active registrations which may interfere with deletion: [crn:v1:bluemix:public:is:eu-es-1:a/abac0df06b644a9cabc6e44f55b3880e::volume:r050-554b16f2-f1c2-410f-a532-cfe107b33c22 crn:v1:bluemix:public:is:eu-es-1:a/bab556e1c47446ef8da61e399343a3e7::volume:r050-417487be-c7c6-4aa9-82f5-fc6aef97f33f crn:v1:bluemix:public:is:eu-es-1:a/bab556e1c47446ef8da61e399343a3e7::volume:r050-8634fe6d-392a-46c2-a6d9-a469e5f356ee crn:v1:bluemix:public:is:eu-es-1:a/bab556e1c47446ef8da61e399343a3e7::volume:r050-ff361a54-fad6-4735-b0d3-c0f3837aff8b crn:v1:bluemix:public:is:eu-es-2:a/abac0df06b644a9cabc6e44f55b3880e::volume:r050-2f9ca121-20f4-4c50-b9d3-2d9d6a86b78d crn:v1:bluemix:public:is:eu-es-2:a/bab556e1c47446ef8da61e399343a3e7::volume:r050-cd1ed004-dce8-4762-b22e-fc75d3cab348 crn:v1:bluemix:public:is:eu-es-2:a/bab556e1c47446ef8da61e399343a3e7::volume:r050-d0e1dcc1-85f9-4e55-931d-fbe71366a5e8 crn:v1:bluemix:public:is:eu-es-2:a/bab556e1c47446ef8da61e399343a3e7::volume:r050-ec590829-ed82-45d2-9e2b-d8f6fbf401f0 crn:v1:bluemix:public:is:eu-es-3:a/abac0df06b644a9cabc6e44f55b3880e::volume:r050-1c0ac71d-52c9-4345-87b7-d2007ace0d66 crn:v1:bluemix:public:is:eu-es-3:a/bab556e1c47446ef8da61e399343a3e7::volume:r050-2b08a94a-7077-4e26-bbde-c18cfd6d6b16 crn:v1:bluemix:public:is:eu-es-3:a/bab556e1c47446ef8da61e399343a3e7::volume:r050-3e68fe7e-a228-4b0d-99a5-a19da5193b1e crn:v1:bluemix:public:is:eu-es-3:a/bab556e1c47446ef8da61e399343a3e7::volume:r050-87c63f1b-8a33-4890-a5ce-071218e7e165]
 2024/10/27 06:58:23 Terraform destroy | 
 2024/10/27 06:58:23 Terraform destroy | ---
 2024/10/27 06:58:23 Terraform destroy | id: terraform-a05cda1f
 2024/10/27 06:58:23 Terraform destroy | summary: '[ERROR] Error while deleting: kp.Error:
 2024/10/27 06:58:23 Terraform destroy | correlation_id=''5edcc499-2465-4a70-afbe-8e4ebc46484c'',
 2024/10/27 06:58:23 Terraform destroy |   msg=''Conflict: Key could not be deleted: Please see `reasons` for more details
 2024/10/27 06:58:23 Terraform destroy |   (PROTECTED_RESOURCE_ERR)'', reasons=''[PROTECTED_RESOURCE_ERR: Key is protecting
 2024/10/27 06:58:23 Terraform destroy |   one or more cloud resources - FOR_MORE_INFO_REFER: https://cloud.ibm.com/docs/key-protect?topic=key-protect-troubleshooting#unable-to-delete-keys]''.
 2024/10/27 06:58:23 Terraform destroy |   The key has the following active registrations which may interfere with deletion:
 2024/10/27 06:58:23 Terraform destroy |   [crn:v1:bluemix:public:is:eu-es-1:a/abac0df06b644a9cabc6e44f55b3880e::volume:r050-554b16f2-f1c2-410f-a532-cfe107b33c22
 2024/10/27 06:58:23 Terraform destroy |   crn:v1:bluemix:public:is:eu-es-1:a/bab556e1c47446ef8da61e399343a3e7::volume:r050-417487be-c7c6-4aa9-82f5-fc6aef97f33f
 2024/10/27 06:58:23 Terraform destroy |   crn:v1:bluemix:public:is:eu-es-1:a/bab556e1c47446ef8da61e399343a3e7::volume:r050-8634fe6d-392a-46c2-a6d9-a469e5f356ee
 2024/10/27 06:58:23 Terraform destroy |   crn:v1:bluemix:public:is:eu-es-1:a/bab556e1c47446ef8da61e399343a3e7::volume:r050-ff361a54-fad6-4735-b0d3-c0f3837aff8b
 2024/10/27 06:58:23 Terraform destroy |   crn:v1:bluemix:public:is:eu-es-2:a/abac0df06b644a9cabc6e44f55b3880e::volume:r050-2f9ca121-20f4-4c50-b9d3-2d9d6a86b78d
 2024/10/27 06:58:23 Terraform destroy |   crn:v1:bluemix:public:is:eu-es-2:a/bab556e1c47446ef8da61e399343a3e7::volume:r050-cd1ed004-dce8-4762-b22e-fc75d3cab348
 2024/10/27 06:58:23 Terraform destroy |   crn:v1:bluemix:public:is:eu-es-2:a/bab556e1c47446ef8da61e399343a3e7::volume:r050-d0e1dcc1-85f9-4e55-931d-fbe71366a5e8
 2024/10/27 06:58:23 Terraform destroy |   crn:v1:bluemix:public:is:eu-es-2:a/bab556e1c47446ef8da61e399343a3e7::volume:r050-ec590829-ed82-45d2-9e2b-d8f6fbf401f0
 2024/10/27 06:58:23 Terraform destroy |   crn:v1:bluemix:public:is:eu-es-3:a/abac0df06b644a9cabc6e44f55b3880e::volume:r050-1c0ac71d-52c9-4345-87b7-d2007ace0d66
 2024/10/27 06:58:23 Terraform destroy |   crn:v1:bluemix:public:is:eu-es-3:a/bab556e1c47446ef8da61e399343a3e7::volume:r050-2b08a94a-7077-4e26-bbde-c18cfd6d6b16
 2024/10/27 06:58:23 Terraform destroy |   crn:v1:bluemix:public:is:eu-es-3:a/bab556e1c47446ef8da61e399343a3e7::volume:r050-3e68fe7e-a228-4b0d-99a5-a19da5193b1e
 2024/10/27 06:58:23 Terraform destroy |   crn:v1:bluemix:public:is:eu-es-3:a/bab556e1c47446ef8da61e399343a3e7::volume:r050-87c63f1b-8a33-4890-a5ce-071218e7e165]'
 2024/10/27 06:58:23 Terraform destroy | severity: error
 2024/10/27 06:58:23 Terraform destroy | resource: ibm_kms_key
 2024/10/27 06:58:23 Terraform destroy | operation: delete
 2024/10/27 06:58:23 Terraform destroy | component:
 2024/10/27 06:58:23 Terraform destroy |   name: github.com/IBM-Cloud/terraform-provider-ibm
 2024/10/27 06:58:23 Terraform destroy |   version: 1.70.1
 2024/10/27 06:58:23 Terraform destroy | ---
```

**Comment by Ak-sky** [comment]
PR was failed due to - `The network zone '0aef01eb77b1921b03d1e2de30d84bf6' is referenced by one or more rules and cannot be deleted.`. This is a known provider [issue](https://github.com/IBM-Cloud/terraform-provider-ibm/issues/4997) for this.

```
2024/10/28 07:49:14 Terraform destroy | Error: DeleteZoneWithContext failed The network zone '0aef01eb77b1921b03d1e2de30d84bf6' is referenced by one or more rules and cannot be deleted.
 2024/10/28 07:49:14 Terraform destroy | {
 2024/10/28 07:49:14 Terraform destroy |     "StatusCode": 412,
 2024/10/28 07:49:14 Terraform destroy |     "Headers": {
 2024/10/28 07:49:14 Terraform destroy |         "Akamai-Grn": [
 2024/10/28 07:49:14 Terraform destroy |             "0.ef2a0660.1730101237.c8534af"
 2024/10/28 07:49:14 Terraform destroy |         ],
 2024/10/28 07:49:14 Terraform destroy |         "Cache-Control": [
 2024/10/28 07:49:14 Terraform destroy |             "no-cache,no-store"
 2024/10/28 07:49:14 Terraform destroy |         ],
 2024/10/28 07:49:14 Terraform destroy |         "Connection": [
 2024/10/28 07:49:14 Terraform destroy |             "keep-alive"
 2024/10/28 07:49:14 Terraform destroy |         ],
 2024/10/28 07:49:14 Terraform destroy |         "Content-Length": [
 2024/10/28 07:49:14 Terraform destroy |             "228"
 2024/10/28 07:49:14 Terraform destroy |         ],
 2024/10/28 07:49:14 Terraform destroy |         "Content-Type": [
 2024/10/28 07:49:14 Terraform destroy |             "application/json; charset=utf-8"
 2024/10/28 07:49:14 Terraform destroy |         ],
 2024/10/28 07:49:14 Terraform destroy |         "Date": [
 2024/10/28 07:49:14 Terraform destroy |             "Mon, 28 Oct 2024 07:40:37 GMT"
 2024/10/28 07:49:14 Terraform destroy |         ],
 2024/10/28 07:49:14 Terraform destroy |         "Expires": [
 2024/10/28 07:49:14 Terraform destroy |             "Thursday, 1 January 1970 00:00:00 GMT"
 2024/10/28 07:49:14 Terraform destroy |         ],
 2024/10/28 07:49:14 Terraform destroy |         "Pragma": [
 2024/10/28 07:49:14 Terraform destroy |             "no-cache"
 2024/10/28 07:49:14 Terraform destroy |         ],
 2024/10/28 07:49:14 Terraform destroy |         "Strict-Transport-Security": [
 2024/10/28 07:49:14 Terraform destroy |             "max-age=31536000; includeSubDomains"
 2024/10/28 07:49:14 Terraform destroy |         ],
 2024/10/28 07:49:14 Terraform destroy |         "Transaction-Id": [
 2024/10/28 07:49:14 Terraform destroy |             "4d4b7a0a-6f11-4123-a00a-19f088afd68e"
 2024/10/28 07:49:14 Terraform destroy |         ],
 2024/10/28 07:49:14 Terraform destroy |         "X-Content-Type-Options": [
 2024/10/28 07:49:14 Terraform destroy |             "nosniff"
 2024/10/28 07:49:14 Terraform destroy |         ],
 2024/10/28 07:49:14 Terraform destroy |         "X-Correlation-Id": [
 2024/10/28 07:49:14 Terraform destroy |             "4d4b7a0a-6f11-4123-a00a-19f088afd68e"
 2024/10/28 07:49:14 Terraform destroy |         ],
 2024/10/28 07:49:14 Terraform destroy |         "X-Proxy-Upstream-Service-Time": [
 2024/10/28 07:49:14 Terraform destroy |             "265"
 2024/10/28 07:49:14 Terraform destroy |         ],
 2024/10/28 07:49:14 Terraform destroy |         "X-Ratelimit-Limit": [
 2024/10/28 07:49:14 Terraform destroy |             "10"
 2024/10/28 07:49:14 Terraform destroy |         ],
 2024/10/28 07:49:14 Terraform destroy |         "X-Ratelimit-Remaining": [
 2024/10/28 07:49:14 Terraform destroy |             "9"
 2024/10/28 07:49:14 Terraform destroy |         ],
 2024/10/28 07:49:14 Terraform destroy |         "X-Ratelimit-Reset": [
 2024/10/28 07:49:14 Terraform destroy |             "1730101238"
 2024/10/28 07:49:14 Terraform destroy |         ]
 2024/10/28 07:49:14 Terraform destroy |     },
 2024/10/28 07:49:14 Terraform destroy |     "Result": {
 2024/10/28 07:49:14 Terraform destroy |         "errors": [
 2024/10/28 07:49:14 Terraform destroy |             {
 2024/10/28 07:49:14 Terraform destroy |                 "code": "zone_used_by_rule",
 2024/10/28 07:49:14 Terraform destroy |                 "message": "The network zone '0aef01eb77b1921b03d1e2de30d84bf6' is referenced by one or more rules and cannot be deleted."
 2024/10/28 07:49:14 Terraform destroy |             }
 2024/10/28 07:49:14 Terraform destroy |         ],
```

**Comment by Ak-sky** [comment]
Upon further investigation, I could see the zone still exists but not the rule, and the likely cause for this failure is delay in deletion of the CBR rule and then parallel deletion of the zone kicked in without checking if the corresponding rule got deleted or not.
image

**Comment by Ak-sky** [comment]
Failed with the same error as this [one](https://github.com/terraform-ibm-modules/terraform-ibm-base-ocp-vpc/pull/529#issuecomment-2439910415) `Key is protecting one or more cloud resources`.

> Error: [ERROR] Error while deleting: kp.Error: correlation_id='52c602b9-a92d-4d90-91e9-0a9116e4579f', msg='Conflict: Key could not be deleted: Please see `reasons` for more details (PROTECTED_RESOURCE_ERR)', reasons='[PROTECTED_RESOURCE_ERR: Key is protecting one or more cloud resources -

**Review by ocofaigh — APPROVED** [review]

**Pull request merged**